File: app/src/main/java/hu/vmiklos/plees_tracker/SleepsAdapter.java
/*
 * Copyright 2019 Miklos Vajna. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

package hu.vmiklos.plees_tracker;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * This is the adapter between RecyclerView and SleepDao.
 */
public class SleepsAdapter
    extends RecyclerView.Adapter<SleepsAdapter.SleepViewHolder>
{
    private List<Sleep> mData;
    private LayoutInflater mLayoutInflater;

    public SleepsAdapter(Context context)
    {
        mData = new ArrayList<>();
        mLayoutInflater = (LayoutInflater)context.getSystemService(
            Context.LAYOUT_INFLATER_SERVICE);
    }

    @NonNull
    @Override
    public SleepViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
                                              int viewType)
    {
        View view =
            mLayoutInflater.inflate(R.layout.layout_sleep_item, parent, false);
        return new SleepViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull SleepViewHolder holder, int position)
    {
        Sleep sleep = mData.get(position);
        holder.start.setText(DataModel.formatTimestamp(new Date(sleep.start)));
        holder.stop.setText(DataModel.formatTimestamp(new Date(sleep.stop)));
        long durationMS = sleep.stop - sleep.start;
        String durationText = DataModel.formatDuration(durationMS / 1000);
        holder.duration.setText(durationText);
    }

    @Override public int getItemCount() { return mData.size(); }

    public void setData(List<Sleep> data)
    {
        if (mData == null)
        {
            mData = data;
            return;
        }

        SleepDiffCallback sleepDiffCallback =
            new SleepDiffCallback(mData, data);
        DiffUtil.DiffResult diffResult =
            DiffUtil.calculateDiff(sleepDiffCallback);
        mData.clear();
        mData.addAll(data);
        diffResult.dispatchUpdatesTo(this);
    }

    /**
     * The view holder holds all views that will display one Sleep.
     */
    class SleepViewHolder extends RecyclerView.ViewHolder
    {
        public final TextView start;
        public final TextView stop;
        public final TextView duration;

        public SleepViewHolder(View view)
        {
            super(view);
            start = view.findViewById(R.id.sleep_item_start);
            stop = view.findViewById(R.id.sleep_item_stop);
            duration = view.findViewById(R.id.sleep_item_duration);
        }
    }

    class SleepDiffCallback extends DiffUtil.Callback
    {
        private final List<Sleep> mOldList;
        private final List<Sleep> mNewList;

        public SleepDiffCallback(List<Sleep> oldList, List<Sleep> newList)
        {
            mOldList = oldList;
            mNewList = newList;
        }

        @Override public int getOldListSize() { return mOldList.size(); }

        @Override public int getNewListSize() { return mNewList.size(); }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition)
        {
            return mOldList.get(oldItemPosition).sid ==
                mNewList.get(newItemPosition).sid;
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition,
                                          int newItemPosition)
        {
            return mOldList.get(oldItemPosition)
                .equals(mNewList.get(newItemPosition));
        }
    }
}

File: app/src/main/java/hu/vmiklos/plees_tracker/SleepsAdapter.kt
/*
 * Copyright 2019 Miklos Vajna. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

package hu.vmiklos.plees_tracker

import android.content.SharedPreferences
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import java.util.ArrayList
import java.util.Date

/**
 * This is the adapter between RecyclerView and SleepDao.
 */
class SleepsAdapter(
    private val viewModel: MainViewModel,
    private val preferences: SharedPreferences
) : RecyclerView.Adapter<SleepsAdapter.SleepViewHolder>() {
    var data: List<Sleep> = ArrayList()
        set(newData) {
            val previousData = field
            field = newData
            DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return previousData.size
                }

                override fun getNewListSize(): Int {
                    return newData.size
                }

                /**
                 * Compares old and new based on their ID only.
                 */
                override fun areItemsTheSame(
                    oldItemPosition: Int,
                    newItemPosition: Int
                ): Boolean {
                    val oldSid = previousData[oldItemPosition].sid
                    val newSid = newData[newItemPosition].sid
                    return oldSid == newSid
                }

                /**
                 * Compares old and new based on their value.
                 */
                override fun areContentsTheSame(
                    oldItemPosition: Int,
                    newItemPosition: Int
                ): Boolean {
                    return previousData[oldItemPosition] == newData[newItemPosition]
                }
            }).dispatchUpdatesTo(this)
        }

    var clickCallback: View.OnClickListener? = null

    override fun getItemCount(): Int {
        return data.size
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): SleepViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.layout_sleep_item, parent, false)
        clickCallback?.let {
            view.setOnClickListener(it)
        }
        return SleepViewHolder(view)
    }

    override fun onBindViewHolder(holder: SleepViewHolder, position: Int) {
        val sleep = data[position]
        holder.start.text = DataModel.formatTimestamp(Date(sleep.start))
        holder.stop.text = DataModel.formatTimestamp(Date(sleep.stop))
        val durationMS = sleep.stop - sleep.start
        val durationText = DataModel.formatDuration(durationMS / 1000)
        holder.duration.text = durationText

        if (preferences.getBoolean("show_awake_for", false)) {
            val nextSleepReferenceTime = if (position == 0) {
                System.currentTimeMillis()
            } else {
                data[position - 1].start
            }
            val durationWakeMS = nextSleepReferenceTime - sleep.stop
            val durationWakeText = DataModel.formatDuration(durationWakeMS / 1000)
            holder.durationWake.text = durationWakeText
        } else {
            holder.durationWakeImage.visibility = View.GONE
            holder.durationWakeHeader.visibility = View.GONE
            holder.durationWakeImage.visibility = View.GONE
        }

        if (preferences.getBoolean("show_rating", false)) {
            holder.rating.rating = sleep.rating.toFloat()
        } else {
            holder.rating.visibility = View.GONE
        }
    }

    /**
     * The view holder holds all views that will display one Sleep.
     */
    inner class SleepViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val start: TextView = view.findViewById(R.id.sleep_item_start)
        val stop: TextView = view.findViewById(R.id.sleep_item_stop)
        val duration: TextView = view.findViewById(R.id.sleep_item_duration)

        val durationWakeImage: View = view.findViewById(R.id.wake_time_image)
        val durationWakeHeader: View = view.findViewById(R.id.wake_time_header)
        val durationWake: TextView = view.findViewById(R.id.wake_item_duration)

        val rating: RatingBar = view.findViewById(R.id.sleep_item_rating)

        val swipeable: View = view.findViewById(R.id.sleep_swipeable)
        val deleteLeft: ImageView = view.findViewById(R.id.sleep_delete_left)
        val deleteRight: ImageView = view.findViewById(R.id.sleep_delete_right)

        fun showSwipeDelete(left: Boolean) {
            deleteLeft.setVisibility(if (left) View.VISIBLE else View.INVISIBLE)
            deleteRight.setVisibility(if (left) View.INVISIBLE else View.VISIBLE)
        }
    }
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */
